third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -62,21 +61,21 @@ enum ValueImpl {
 /// type or not, and a few specialized flavors to offer compact memory layout for small
 /// primitive types.
 ///
-/// Except when not owned by the VM stack, a container always lives inside an Rc<RefCell<>>,
+/// Except when not owned by the VM stack, a container always lives inside an Arc<Mutex<>>,
 /// making it possible to be shared by references.
 #[derive(Debug, Clone)]
 enum Container {
-    Locals(Rc<RefCell<Vec<ValueImpl>>>),
-    Vec(Rc<RefCell<Vec<ValueImpl>>>),
-    Struct(Rc<RefCell<Vec<ValueImpl>>>),
-    VecU8(Rc<RefCell<Vec<u8>>>),
-    VecU64(Rc<RefCell<Vec<u64>>>),
-    VecU128(Rc<RefCell<Vec<u128>>>),
-    VecBool(Rc<RefCell<Vec<bool>>>),
-    VecAddress(Rc<RefCell<Vec<AccountAddress>>>),
-    VecU16(Rc<RefCell<Vec<u16>>>),
-    VecU32(Rc<RefCell<Vec<u32>>>),
-    VecU256(Rc<RefCell<Vec<u256::U256>>>),
+    Locals(Arc<Mutex<Vec<ValueImpl>>>),
+    Vec(Arc<Mutex<Vec<ValueImpl>>>),
+    Struct(Arc<Mutex<Vec<ValueImpl>>>),
+    VecU8(Arc<Mutex<Vec<u8>>>),
+    VecU64(Arc<Mutex<Vec<u64>>>),
+    VecU128(Arc<Mutex<Vec<u128>>>),
+    VecBool(Arc<Mutex<Vec<bool>>>),
+    VecAddress(Arc<Mutex<Vec<AccountAddress>>>),
+    VecU16(Arc<Mutex<Vec<u16>>>),
+    VecU32(Arc<Mutex<Vec<u32>>>),
+    VecU256(Arc<Mutex<Vec<u256::U256>>>),
```

Comment:
I'm also very concerned about this change. Are we now allowing multiple VM instances to modify the same value simultaneously, without any copy-on-write semantics?

Reply:
We copy on every resource load into `TransactionDataCache` to avoid modifications of the same value (and similarly we can add some restrictions on `Resource` APIs to avoid copies), but this is checked for the top-level Value only which is ugly...

Reply:
Honestly this flies against all my intuition about making the system performant. Trading copying for contention is almost always a bad trade off. All performance work on google multithreaded servers almost always went the opposite direction, removing shared data even at the cost of copying.

Reply:
Also this seems like an easy path to deadlock hell. Two threads grabbing a pair of mutexes in different order, at least it's not clear to me why that wouldn't happen

Reply:
We should, and here I use `Arc<Mutex<Value>>` exactly: the problem is that having Rcs and RefCells inside stops being safe then, in particular when the storage is shared between concurrent sessions

Reply:
@gerben-stavenga as Zekun mentioned below we trade of not just copying but also (de-)serialization, let me update the PR to reflect this better

Reply:
@georgemitenkov I know but I still think, the trade is wrong. Typically it benefits performance to do extra copying/serialization to ensure that the cores do not have to bang on the same memory. Making everything mutex means the cores are banging on the same memory locations serializing mutex access.

Reply:
I do agree that this looks like anti-pattern, though memcopy is much more efficient than serde roundtrips to bytes. what I imagine is that we could hold immutable data copy as Arc<Value> and convert it to Rc<RefCell<Value>> during actual txn session, haven't read through the code, so not easy to tell how hard it would be.

Reply:
Oh I agree with that, at the end points of the computation (start values and result values) we could do something faster (ie. memcpy-ish).